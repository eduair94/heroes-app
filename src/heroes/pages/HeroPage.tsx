import { Navigate, useNavigate, useParams } from "react-router-dom";
import { getHeroById } from "../helpers";
import { useMemo } from "react";
import { baseUrl } from "../../helpers";

export const HeroPage = () => {
  const { id } = useParams();

  const hero = useMemo(() => getHeroById(id as string), [id]);

  const navigate = useNavigate();

  const onNavigateBack = () => {
    navigate(-1);
  };

  if (!hero) {
    return <Navigate to="/" />;
  }

  return (
    <div className="row mt-5">
      <div className="col-4">
        <img
          src={`${baseUrl}/assets/heroes/${id}.jpg`}
          alt={hero.superhero}
          className="img-thumbnail animate__animated animate__fadeInLeft"
        />
      </div>
      <div className="col-8">
        <h3>{hero.superhero}</h3>
        <ul className="list-group list-group-flush">
          <li className="list-group-item">
            <b>Alert ego:</b> {hero.alter_ego}
          </li>
          <li className="list-group-item">
            <b>Publisher:</b> {hero.publisher}
          </li>
          <li className="list-group-item">
            <b>First appearence:</b> {hero.first_appearance}
          </li>
        </ul>
        <h5 className="mt-3">Characters</h5>
        <p>{hero.characters}</p>

        <button onClick={onNavigateBack} className="btn btn-outline-primary">
          Back
        </button>
      </div>
    </div>
  );
};
