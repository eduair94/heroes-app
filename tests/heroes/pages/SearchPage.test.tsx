import { fireEvent, render, screen } from "@testing-library/react";
import React from "react";
import { MemoryRouter } from "react-router-dom";
import { SearchPage } from "../../../src/heroes/pages/SearchPage";

jest.mock("../../../src/helpers/getBaseUrl", () => ({
  baseUrl: "",
}));

const mockedUseNavigate = jest.fn();
jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => mockedUseNavigate,
}));

describe("Tests in <SearchPage/>", () => {
  test("should display properly with default values", () => {
    const { container } = render(
      <MemoryRouter>
        <SearchPage />
      </MemoryRouter>,
    );
    expect(container).toMatchSnapshot();
  });

  test("It must show Batman and the input with the value queryString", () => {
    render(
      <MemoryRouter initialEntries={["/search?q=batman"]}>
        <SearchPage />
      </MemoryRouter>,
    );
    const input = screen.getByRole("textbox") as HTMLInputElement;
    expect(input.value).toBe("batman");

    const img: HTMLImageElement = screen.getByRole("img");
    expect(img.src).toContain("batman");

    const error = screen.getByLabelText("alert-danger");
    expect(error.style.display).toBe("none");
  });

  test("It must show an error if the hero does not exist", () => {
    render(
      <MemoryRouter initialEntries={["/search?q=non_existant"]}>
        <SearchPage />
      </MemoryRouter>,
    );
    const input = screen.getByRole("textbox") as HTMLInputElement;
    expect(input.value).toBe("non_existant");

    const error = screen.getByLabelText("alert-danger");
    expect(error.style.display).not.toBe("none");
  });

  test("It must call navigate in the new screen", () => {
    const searchText = "superman";
    render(
      <MemoryRouter initialEntries={["/search"]}>
        <SearchPage />
      </MemoryRouter>,
    );
    const input = screen.getByRole("textbox") as HTMLInputElement;
    const form = screen.getByLabelText("form-hero");
    fireEvent.input(input, {
      target: { name: "searchText", value: searchText },
    });
    fireEvent.submit(form);

    expect(mockedUseNavigate).toHaveBeenCalledWith(`?q=${searchText}`);
  });
});
