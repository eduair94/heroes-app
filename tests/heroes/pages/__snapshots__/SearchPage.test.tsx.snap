// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in <SearchPage/> should display properly with default values 1`] = `
<div>
  <h1>
    Search
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-5"
    >
      <h4>
        Searching
      </h4>
      <hr />
      <form
        aria-label="form-hero"
      >
        <input
          autocomplete="off"
          class="form-control"
          name="searchText"
          placeholder="Search a hero"
          type="text"
          value=""
        />
        <button
          class="btn btn-outline-primary mt-2"
        >
          Search
        </button>
      </form>
    </div>
    <div
      class="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        aria-label="search-hero"
        class="alert alert-primary animate__animated animate__fadeIn"
      >
        Search a hero
      </div>
      <div
        aria-label="alert-danger"
        class="alert alert-danger animate__animated animate__fadeIn"
        style="display: none;"
      >
        No hero with 
        <b />
      </div>
      <div
        class="row gy-3 row-cols-1 g-3"
      />
    </div>
  </div>
</div>
`;
